tests/test_pyo3test.rs

```diff
@@ -0,0 +1,74 @@
+#![cfg(not(any(PyPy, GraalPy, Py_3_7, Py_3_8, all(windows, Py_LIMITED_API, Py_3_9))))]
+use pyo3::{prelude::*, types::PyDict};
+
+// The example from the Guide ...
+fn o3_addone(num: isize) -> isize {
+    num + 1
+}
+
+#[pyfunction]
+#[pyo3(name = "addone")]
+fn py_addone(num: isize) -> isize {
+    o3_addone(num)
+}
+
+#[pymodule]
+#[pyo3(name = "adders")]
+fn py_adders(module: &Bound<'_, PyModule>) -> PyResult<()> {
+    module.add_function(wrap_pyfunction!(py_addone, module)?)?;
+    Ok(())
+}
+
+// This is how the test would be written WITHOUT using the pyo3test macro. This validates that
+// adders.addone is correctly constructed.
+#[test]
+fn test_pyo3test_without_macro() {
+    pyo3::prepare_freethreaded_python();
+    Python::with_gil(|py| {
+        let sys = PyModule::import_bound(py, "sys").unwrap();
+        let py_modules: Bound<'_, PyDict> =
+            sys.getattr("modules").unwrap().downcast_into().unwrap();
+        let py_adders_pymodule = unsafe { Bound::from_owned_ptr(py, py_adders::__pyo3_init()) };
```

Comment:
This is an expanded example of how modules are initialised and exposed in a pre-initialised interpreter. It is based on the example in [the Guide section 3.4](https://pyo3.rs/main/python-from-rust/calling-existing-code#want-to-embed-python-in-rust-with-additional-modules):

1. `sys` is imported and `sys.modules` assigned to a variable
1. the module is initialised
1. the module is inserted into `sys.modules`

I believe that the lifetime of py_adders_pymodule is safe as:

- `py_adders::__pyo3_init()` is called with the GIL lock in place.
- Therefore the GIL handler will hand out the same `'py` lifetime as the rest of the code is using.
- `Bound::from_owned_ptr()` is passed the `py` token from this GIL lock, binding the initialised module to the correct lock
- and the whole process is performed as a single action.

Have I correctly understood how the GIL lock is handled w.r.t. the lifetime of the `PyModule` created by `__pyo3_init`?

`append_to_inittab!` is not possible here, because there is no guarantee that the interpreter is not already initialised.

This sadly also means that compatibility is limited to cPython>=3.9 if I understand [make_module](https://github.com/PyO3/pyo3/blob/f5fee94afcaf11edceceed45192aabd71aeb9415/src/impl_/pymodule.rs#L127) correctly.